### PR TITLE
feat!: migrate to vue 3

### PR DESCRIPTION
**Pull request by emuvente**
Initial PR for migrating from webpack & vue 2 to vite & vue 3. This still has many issues but the overall functionality is there. I've had to manually roll-back changes while working on this and change some parts multiple times, so there are some minor related issues that still need to be cleaned up. The general process for developing this has been cycling through running `npm run build`, `npm run lint`, `npm run unit`, `npm run storybook`, `npm run dev -- --config=local`, and `npm start -- --config=local`, and fixing any errors that arise.

## References

https://vueschool.io/articles/vuejs-tutorials/how-to-migrate-from-vue-cli-to-vite/
https://www.sitepoint.com/webpack-vite-migration/
https://vitejs.dev/guide/ssr
https://storybook.js.org/docs/migration-guide/from-older-version

## Tasks completed

- upgraded to node 20
- many dependency upgrades (see last section)
- switched to `"type": "module"`
- replaced require.context with import.meta.glob and importHelpers util
- used full paths for css imports
- replaced all instances of require with import
- added `defineAsyncComponent` for all async component imports
- replaced `>>>` with `:deep()`
- replaced `/` with `math.div()` in sass & scss
- replaced `import '@/...'` with `import '#src/...'` and removed `@` alias
- removed `algolia` and related components
- converted server modules to esm where possible
- removed/migrated vue filters
- swapped `workerpool` for `piscina`
- changed the imports for { gql } from @apollo/client to graphql-tag
- added babel transformer to handle import.meta.url in jest
- updated jest svg transform
- changed `updateProps` to `rerender` in specs

## TODO

- [x] preload hints for static assets
- [x] fix/complete static asset url resolution
- [x] complete production ssr
- [x] review need for plugin-legacy with current browser support target
- [x] upgrade cypress
- [x] fix failing jest tests
- [x] fix individual page issues
- [x] performance testing
- [x] fix failing stories
- [x] update manifest (`manifest.webmanifest`) to have full URL `src` properties -> warnings can be seen after running `npm run build` and `npm start -- --config=local`
- [x] ensure production minification of HTML and CSS (it happened before with webpack) -> see commented out `htmlPurge` in `vite.config.js`

## Items to review/QA

These are things that have changed in this PR that need review/QA, and issues that have popped up while developing that still need to be addressed.

- [x] focus lock issues when navigating on /15
- [x] KvLightbox vue focus lock div
- [x] css preprocessing
- [x] font loading
- [x] flag icons
- [x] icon sprite sheet loading
- [x] vue sfc deep selector usage
- [x] live loans loan use and loan callouts
- [x] local graphql resolvers default values
- [x] qr code for mfa
- [x] KvLoadingSpinner
- [x] DynamicRichText
- [x] auto deposit signup form

## Setup local development

Note: Your terminal app must be running with Rosetta for backwards compatibility with some of our dependencies, `canvas` in particular.

```
# Terminal window 1
npm checkout vue-3
nvm use
rm -rf node_modules
npm i
npm run dev -- --config=dev-custom-host

# Terminal window 2
caddy run

# Open custom host in browser, for example https://kiva-ui.local/lend/filter
```

### CODE REVIEW

**Comment by eddieferrer** [comment]
Comment on my latest commit: I had to change the imports for { gql } from @apollo/client to graphql-tag to get the ESM support, lots of discussion on this, see: https://github.com/apollographql/apollo-feature-requests/issues/287

**Comment by eddieferrer** [comment]
More work on Jest testing complete: 
These specs are now passing: 
test/unit/specs/components/WwwFrame/TheTipMessage.spec.js 
test/unit/specs/components/WwwFrame/TheHeader.spec.js 
test/unit/specs/components/Checkout/BasketItem.spec.js 
test/unit/specs/components/WwwFrame/SearchBar.spec.js

I had to comment out a test in `TheTipMessage.spec.js ` and have added a comment.

**Comment by eddieferrer** [comment]
More fixes related to small component changes and unit testing specs. 
Now at just 6 failing specs.

**Comment by eddieferrer** [comment]
@emuvente These are the only unit tests that are failing at the moment: 

`test/unit/specs/components/Kv/KvCheckboxList.spec.js` -- I think this component might actually be broken and the test is fine, it would be helpful to have storybook running to see this component

`test/unit/specs/util/getDeepComponents.spec.js` -- I'm hoping that you know how to fix this one after your changes to the asyncComponent definitions.

**Comment by emuvente** [comment]
Storybook loads now, though there are issues with individual stories. It seems like the format for passing in arguments as props may have changed.

I started looking into improving `getDeepComponents` but I found that it might not work at all with the composition API. I think we may need to change how we do data prefetching for vue 3.

**Comment by emuvente** [comment]
Ok, just pushed up a fix for `getDeepComponents`, so there only remaining unit test failure is `KvCheckboxList`.

Prefetching is in an unstable state. It currently will only work as long as all the components are using the options api. For any component that uses the composition api they will not have a prefetch option and none of the prefetches will be called for any of their child components.

I've looked into the various ways we can do prefetching in vue 3, and there isn't any easy solution. There is a super easy and useful [serverPrefetch hook](https://vuejs.org/api/composition-api-lifecycle.html#onserverprefetch), but it only gets called during SSR, so client-side prefetching before navigation isn't supported. For the client-side we would need to use a combination of [beforeRouteEnter and beforeRouteUpdate](https://router.vuejs.org/guide/advanced/data-fetching.html#Fetching-Before-Navigation). So to do the same sort of prefetching we are doing today, we would need to define the data fetch in 3 different hooks in each component, each hook using a slightly different format.

I would recommend that we try to replace prefetching with loading data after page navigation. This will require rewriting some components and adding loading states for many pages (blank areas are ok, no need for loading placeholders/spinners everywhere as long as content shifting is avoided). On the plus side it will mean much faster performance for server renders and better perceived client-side performance.

**Comment by dyersituations** [comment]
All unit tests now pass: I quieted the `console.warn` messages from `live-loan-fetch.spec.js` and fixed a functional issue with `KvCheckboxList`.

It looks like story requirements have indeed changed a bit, and a simplified story definition now works. Here's an example from `KvCheckboxList`. I'll look through the rest of the stories next and resolve similar issues:

```
import KvCheckboxList from '#src/components/Kv/KvCheckboxList';

export default {
	title: 'Kv/Form Elements/KvCheckboxList',
	component: KvCheckboxList,
};

const items = [...Array(4)].map((_, i) => ({ value: `${i}`, title: `Option ${i}` }));

export const Default = { args: { items } };

export const ShowSelectAll = { args: { items, showSelectAll: true } };
```

**Comment by dyersituations** [comment]
In case anyone else encounters issues with setup, here's what I had to do:

- Install vips (this takes a couple hours): `brew install vips`
- Ensure VS Code uses Rosetta in console. Open settings JSON and add:
```
    "terminal.integrated.profiles.osx": {
        "x86 zsh": {
            "path": "/usr/bin/arch",
            "args": ["-arch", "x86_64", "/bin/zsh"]
        }
    },
    "terminal.integrated.defaultProfile.osx": "x86 zsh"
```
- Restart VS Code and open new terminal
- Remove node_modules: `rm -rf node_modules`
- Use/install correct node version: `nvm use` (I made sure to install after changing Rosetta version)
- Install packages again: `npm i`

At this point both `npm run unit` and `npm run storybook` worked. When opening a new terminal make sure to run `nvm use` if the terminal default isn't the right version of node.

**Comment by dyersituations** [comment]
Another story format that works in Storybook 8 when we need to add mixins. The main difference here is ensuring that we return the `args` in the `setup` method:

```
const args = {
	goal: 50,
	progress: 25,
};

export default {
	title: 'Components/Checkout/DepositIncentiveUpsell',
	component: DepositIncentiveUpsell,
	args,
};

export const Default = (args, { argTypes }) => ({
	props: Object.keys(argTypes),
	components: { DepositIncentiveUpsell },
	mixins: [apolloStoryMixin({ queryResult })],
	setup() { return args; },
	template: '<deposit-incentive-upsell :goal="goal" :progress="progress" />',
});
```

**Comment by dyersituations** [comment]
I've now gone through all of the stories and encountered a number of trends in issues. The quick to fix issues were resolved, but these new categories of errors need to be investigated.

- font doesn’t match deployed storybook from before migration -> new fonts missing?
- DonationNudgeLightbox -> console error
- EcoChallengeLightbox -> $filters error/warning
- LoanSpotlight -> prefetch issue
- MonthlyGoodSelector -> process is not defined (vuelidate)
- MonthlyGoodUpdateForm -> process is not defined (vuelidate)
- MultiAmountSelector -> process is not defined (vuelidate)
- KvCurrencyInput -> process is not defined (vuelidate)
- KvPhoneInput -> warning about $listeners, missing images
- KvRadio -> warning about $listeners
- KvSelect -> warning about $listeners
- KvVerificationCodeInput -> warning about $listeners
- KvButton -> createElement is not a function
- KvCarousel -> require is not defined
- KvDynamicHeroClassic -> createElement is not a function
- KvExpandableQuestion -> requested module does not provide an export named ‘default’
- KvFlag -> some SVGs don’t render, some stories encounter “Cannot read properties of undefined (reading 'toUpperCase’)”
- KvFrequentlyAskedQuestions -> $filters error, “Cannot read properties of undefined (reading 'changeCase’)”
- KvHero -> require is not defined
- KvIcon -> require is not defined
- KvIconButton -> createElement is not a function
- KvLightbox -> createElement is not a function
- KvProgressBar -> t.toFixed is not a function
- GridLoanCard -> requested module does not provide an export named ‘default’
- ListLoanCard -> Cannot read properties of undefined (reading 'numeral')
- KvLoanCards (New Home Page Loan Card) -> $filters warning and “Cannot read properties of undefined (reading 'numeral’)”
- RecommendedLoanCard -> $filters warning and “Cannot read properties of undefined (reading 'numeral’)”
- Challenge Header (Loan Search) -> warning about missing goal prop, might be because component name is identical to another that has that prop
- KivaMultiCategoryGrid -> Cannot read properties of undefined (reading 'right') -> embla-carousel
- 15Years -> Failed to fetch dynamically imported module
- BorrowerProfile -> Failed to fetch dynamically imported module
- CorporateCampaignLanding -> Failed to fetch dynamically imported module
- GetStarted -> requested module does not provide an export named ‘default’
- KivaCreditStats -> Cannot read properties of undefined (reading 'numeral')
- SecuritySettings -> The requested module does not provide an export named 'default'
- ThanksPage -> The requested module does not provide an export named 'default'
- AppealBannerCircular -> The requested module does not provide an export named 'default'
- DepositIncentiveBanner -> The requested module does not provide an export named 'default'
- DonationBanner -> The requested module does not provide an export named 'default’
- GenericPromoBanner -> The requested module does not provide an export named 'default'
- PromoCreditBanner -> The requested module does not provide an export named 'default'
- TheFooterCorporate -> require is not defined
- TheHeader -> The requested module does not provide an export named 'default'
- WwwPage -> Cannot read properties of undefined (reading 'numeral')
- WwwPageCorporate -> require is not defined
- WwwPageMinimal -> createElement is not a function
- A bunch of files that can’t be found (messages appear in terminal running storybook)
Neither '#src/plugins/loan-group-categories.vue' nor '#src/plugins/loan-group-categories.js(x)' or '#src/plugins/loan-group-categories/index.js(x)' or '#src/plugins/loan-group-categories/index.ts(x)' could be found in '/Users/caseyd/kiva/ui/src/components/MonthlyGood'
Neither '#src/plugins/click-outside.vue' nor '#src/plugins/click-outside.js(x)' or '#src/plugins/click-outside/index.js(x)' or '#src/plugins/click-outside/index.ts(x)' could be found in '/Users/caseyd/kiva/ui/src/components/MonthlyGood'
Neither '#src/plugins/team-goal-mixin.vue' nor '#src/plugins/team-goal-mixin.js(x)' or '#src/plugins/team-goal-mixin/index.js(x)' or '#src/plugins/team-goal-mixin/index.ts(x)' could be found in '/Users/caseyd/kiva/ui/src/components/Thanks'
Neither '#src/plugins/loan-channel-query-map.vue' nor '#src/plugins/loan-channel-query-map.js(x)' or '#src/plugins/loan-channel-query-map/index.js(x)' or '#src/plugins/loan-channel-query-map/index.ts(x)' could be found in '/Users/caseyd/kiva/ui/src/components/LoanCollections'
Neither '#src/plugins/loan/percent-raised-mixin.vue' nor '#src/plugins/loan/percent-raised-mixin.js(x)' or '#src/plugins/loan/percent-raised-mixin/index.js(x)' or '#src/plugins/loan/percent-raised-mixin/index.ts(x)' could be found in '/Users/caseyd/kiva/ui/src/components/LoanCards'
Neither '#src/plugins/loan/time-left-mixin.vue' nor '#src/plugins/loan/time-left-mixin.js(x)' or '#src/plugins/loan/time-left-mixin/index.js(x)' or '#src/plugins/loan/time-left-mixin/index.ts(x)' could be found in '/Users/caseyd/kiva/ui/src/components/LoanCards'
Neither '#src/plugins/add-to-basket-show-interstitial.vue' nor '#src/plugins/add-to-basket-show-interstitial.js(x)' or '#src/plugins/add-to-basket-show-interstitial/index.js(x)' or '#src/plugins/add-to-basket-show-interstitial/index.ts(x)' could be found in '/Users/caseyd/kiva/ui/src/components/LoanCards/Buttons'
Neither '#src/plugins/contentful-ui-setting-styles-mixin.vue' nor '#src/plugins/contentful-ui-setting-styles-mixin.js(x)' or '#src/plugins/contentful-ui-setting-styles-mixin/index.js(x)' or '#src/plugins/contentful-ui-setting-styles-mixin/index.ts(x)' could be found in '/Users/caseyd/kiva/ui/src/components/Contentful'

**Comment by dyersituations** [comment]
Issue with `TheTipMessage` resolved. Turns out that Vue 3 runs the `immediate` watchers before the `$refs` have been defined. Using `$nextTick` was a simple workaround to get the behavior we wanted.

**Comment by dyersituations** [comment]
Today I've been investigating an issue with Vue component CSS pop-in, and I eventually tracked it down to the component CSS being missing from the SSR rendering in local development. In order to remove the component FOUC (Flash of Unstyled Content) in local dev, I basically disabled SSR in local dev but passing in empty string for the `appHtml` only for dev. I verified that SSR still works once the app is built and running locally. You'll now see a warning in local dev (shown below), and a brief white screen on page load, but the CSS no longer pops in, which makes local dev a bit less jolting.

```
Attempting to hydrate existing markup but container is empty
```

**Comment by dyersituations** [comment]
The `lend/filter` page now has all logged-out features functional. We can't currently login, so I can't test saved search saving, but all other filtering functionality is working.

**Comment by dyersituations** [comment]
The category pages now load and all functionality (other than ATB and login) works as expected. In order to get all buttons working, I had to also fix the initial load of the monthly good page. I encountered that the version of `vuelidate` isn't compatible with Vue 3, so I installed the version for Vue 3 and updated the form on the monthly good page. Monthly good functionality still needs fixes (when you click "get started", the next page has an error on load).

Look to `LandingForm.vue` for an example of using the new version of vuelidate for Vue 3.

**Comment by dyersituations** [comment]
Running Caddy locally now works with Vite and Vue 3. We need to run Caddy to support a non-localhost HTTPS connection, due to Auth0 requiring a secure connection even in development. With the most recent commits, you can now run the `dev-custom-host` configuration successfully with this branch and are able to use: log in, add to basket, and HMR still works (configuration changes were needed for Caddy).

Here's how I've been running it locally (two terminal windows in VS Code):
```
caddy run
----
npm run dev -- --config=dev-custom-host
----
Open https://kiva-ui.local/lend/filter
```

**Comment by dyersituations** [comment]
Latest changes allow the `/checkout` page to load without crashing the dev server (changes to `kv-shop`). The current state of the checkout page is that it loads but encounters a vuelidate error. Luckily there is a known pattern in recent commits from the monthly good page, so that's the next step forward here.

**Comment by dyersituations** [comment]
Latest changes from `main` have been merged, adjusted to match new patterns, and merge conflicts resolved.

**Comment by dyersituations** [comment]
Latest changes from main have been merged, adjusted to match new patterns, and merge conflicts resolved. The app runs again locally in this branch, more fixes to come soon during this time-blocked migration work time.

**Comment by dyersituations** [comment]
All forms that use `vuelidate` now use the new pattern and the old `vuelidate` package was removed. Because not all of the pages in this branch currently load, I didn't functionally test the changed forms, but I added a TODO to the PR description to test the forms once the site works completely in this branch.

**Comment by dyersituations** [comment]
This sprint's time-boxed migration is now completed. Here's what has been completed:
- Merged and migrated the 90+ files that have been changed in `main`
- All forms with `vuelidate` have been updated to the new pattern -> most of the forms have not been functionally tested
- Resolved errors that prevented the borrower profile from loading -> the page now loads but the features on the page have not been tested
- Resolved a number of random errors and issues remaining

**Comment by dyersituations** [comment]
Default (non-experiment) thanks page now loads as expected (no more server errors). One new pattern found was applied through other pages that use `route` in `preFetch`.

**Comment by dyersituations** [comment]
Borrower profile features now appear to work like expected. The lend CTA animation was broken, both in formatting and the animation itself (animation migration was also applied to another component that uses `transition`).

**Comment by dyersituations** [comment]
Category pages seem to work like expected now. There was a bug in how classes were added to the responsive image component. The latest change does add some duplicate attributes, but I don't think it will cause any issues in general.

**Comment by dyersituations** [comment]
The loan finding page (`/lend-by-category`) now all works like expected. There was an issue with the slots/templates for the loan carousels.

**Comment by dyersituations** [comment]
Portfolio page now loads and page features work like expected. Here are the fixes:
- Navigation using Vue router wasn't hitting `preFetch` before `created` like expected
- Treemap commonjs import needed to be updated to ensure `this` would be defined in the module when methods called

**Comment by dyersituations** [comment]
Portfolio settings pages now load, components render correctly, and functionality works.

**Comment by dyersituations** [comment]
All story-related issues have been resolved. Fixes sometimes were needed in the components, sometimes needed in the stories. A few stories are broken, but they were broken in `main` as well.

**Comment by dyersituations** [comment]
Vite build has issues with `webmanifest`, related to the hashed image URLs after build. One workaround mentioned in GitHub issues was to utilize the `public` folder. While this means `favicons` will be cached, we never really update them, so I went forward with that solution. I also tried working with `VitePWA`, but it was heavy-handed to just get `webmanifest` support, and there wasn't a clear path with that library to get hashed paths either. With the solution just committed, the `webmanifest` and favicons are in the `public/static` folder, get copied to `dist/static`, and there are no warnings in the console with either local development or when running the compiled app.

**Comment by dyersituations** [comment]
Resolved issues with `KvFlag` that occurred when running the app but not when running Storybook. I wasn't seeing a quick way to support the sprite icons (related to how webpack used to pull in and change paths for SVG files), so I went with the CSS background approach for the flag package we're using. After updating the usage, I tested all locations where we use `KvFlag` and resolved any issues.

The latest batch of changes should _hopefully_ include the remaining fixes for the custom form elements that live in this repo. There were a number of issues with the autolending settings filters in the modal on that page that have been resolved.

**Comment by dyersituations** [comment]
Turns out that PurgeCSS (the commonly used CSS cleanup package that works with Vue) doesn't support safelist entries targeting attribute selectors. Vue uses `data-v` for scoped CSS. PurgeCSS cannot target those. There's an open issue in the PurgeCSS repo from 2022 about this, and it's what I eventually tracked down the issue to be. The specific UI issue I saw related to this is the scoped styles in `KvLendCta` in `kv-elements` (green lend button was missing correct rounded corners). In other words, there's no good support currently unless we fork PurgeCSS.

That all being said, Tailwind v3 now uses PurgeCSS itself, and I compared the minified tailwind CSS with and without using PurgeCSS in postcss as well. The difference for the index CSS file was 299kb vs 316kb (the former being with the additional PurgeCSS ran in Tailwind automatically and ran in postcss).

So, I think for now we can ignore CSS cleanup in our build process. We'll need to remove our custom CSS ourselves when it isn't being used, but I doubt we have enough custom CSS to make a big difference in output, as shown in those KB comparison numbers.

**Comment by dyersituations** [comment]
Minification verified (working) and performance tests ran locally (Lighthouse scores more or less match dev env with Vue 2 running). Next step is deploying to dev and running through functional and Lighthouse tests again: https://kiva.atlassian.net/browse/MP-718

File: src/pages/Portfolio/ImpactDashboard/DistributionGraphs.vue [hunk 1]
```diff
@@ -23,8 +23,6 @@ import { gql } from 'graphql-tag';
 import delayUntilVisibleMixin from '#src/plugins/delay-until-visible-mixin';
 import getCacheKey from '#src/util/getCacheKey';
 import StatsTable from '#src/components/Stats/StatsTable';
-import KvPieChart from '@kiva/kv-components/vue/KvPieChart';
-import KvTreeMapChart from '@kiva/kv-components/vue/KvTreeMapChart';
```
Comment (mcstover) [inline src/pages/Portfolio/ImpactDashboard/DistributionGraphs.vue]: Don't we still need treemap here as it's used by default below? Or maybe the "chart" prop isn't used any longer...

Comment (dyersituations) [reply src/pages/Portfolio/ImpactDashboard/DistributionGraphs.vue]: That prop actually isn't used. I removed the computed value when I removed these imports, but looks like I missed the unused prop.

**Review by mcstover — APPROVED** [review]
I think it's go time for this...

**Pull request merged**